### PR TITLE
Alicloud DNS support

### LINKED ISSUE #367: Support Alicloud
Alicloud is planned to be fully supported by Gardener project.

Some cases need to be considered:
1.  Cloud Controller Manager
Kubernetes has no in-tree cloud provider for Alicloud. Project https://github.com/AliyunContainerService/alicloud-controller-manager provides solution to enable CCM in Alicloud.
2. Dynamic Volume provisioning
Kubernetes dose not support PVC for Alicloud directly. CSI is a standard interface for container orchestration systems (like Kubernetes) to expose arbitrary storage systems to their container workloads. From Kubernetes v1.10,  CSI is in Beta. https://kubernetes.io/docs/concepts/storage/volumes/#csi . Project https://github.com/AliyunContainerService/csi-plugin could be leveraged.
CSI plugin works with followging components:
https://github.com/kubernetes-csi/external-attacher
https://github.com/kubernetes-csi/driver-registrar
https://github.com/kubernetes-csi/external-provisioner
3. Chinese cyber security law must be followed
How deploy IDP? 
4. Network connection issue
Some web sites like google are banned from China. Docker registry is to be self-hosted or leverage registry provided by Alicloud.
5. Use DNS provided by Alicloud
6. CoreOS image provided by Alicloud
Cloud Init is not supported. Only Clould config is supporte(BASH).

There may be many issues not metioned. Any quesitons/suggestions are welcome.

### PR DESCRIPTION
**Pull request by jia-jerry**
**What this PR does / why we need it**:
Support Alicloud DNS
**Which issue(s) this PR fixes**:
Fixes #367 
**Special notes for your reviewer**:
Please be noted that the terraform DSN record resource structure for Alicloud is quite different with other providers like AWS. No hosted zone id needed.
 
**Release note**:

```noteworthy user
Gardener has now DNS supported for Alicloud. Please consider the [example resource manifests](https://github.com/gardener/gardener/tree/master/example) if you want to use it.
```

### CODE REVIEW

**Comment by jia-jerry** [comment]
Hi @rfranzke, 

I have an issue to log on https://concourse.ci.infra.gardener.cloud with following steps:
1. Use term gardener
2. Log in with Github account.

It says ```verification failed```

Would you please refer an owner to help me log on the concourse system?

Thanks a lot!

**Comment by vlerenc** [comment]
Hi @jia-jerry, you have been invited now.

**Comment by jia-jerry** [comment]
Hi @vlerenc ,
Log in issue still exists with message: ```verification failed```

Would you please help me check it?

Thanks a lot.

**Comment by vlerenc** [comment]
Surprising, but I couldn't look deeper, please contact @ccwienk.

**Comment by jia-jerry** [comment]
Hi @vlerenc 

Log in issue solved. I need to **accept** your invitation. I thought I joined automatically after I received your mail. 

Will follow up the build error.

**Comment by jia-jerry** [comment]
The concouse failure of original PR seemed caused by github server issue:
```
Identity added: /tmp/git-resource-private-key (/tmp/git-resource-private-key)
/usr/lib/ruby/gems/2.4.0/gems/octokit-4.8.0/lib/octokit/response/raise_error.rb:16:in `on_complete': POST https://api.github.com/repos/gardener/gardener/statuses/22c596fef03bf8b302e69f1a667429f4ceee76a9: 502 - Server Error (Octokit::BadGateway)
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/response.rb:9:in `block in call'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/response.rb:61:in `on_complete'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/response.rb:8:in `call'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-http-cache-2.0.0/lib/faraday/http_cache.rb:146:in `call!'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-http-cache-2.0.0/lib/faraday/http_cache.rb:115:in `call'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/rack_builder.rb:143:in `build_response'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/connection.rb:387:in `run_request'
	from /usr/lib/ruby/gems/2.4.0/gems/faraday-0.14.0/lib/faraday/connection.rb:175:in `post'
	from /usr/lib/ruby/gems/2.4.0/gems/sawyer-0.8.1/lib/sawyer/agent.rb:94:in `call'
	from /usr/lib/ruby/gems/2.4.0/gems/octokit-4.8.0/lib/octokit/connection.rb:156:in `request'
	from /usr/lib/ruby/gems/2.4.0/gems/octokit-4.8.0/lib/octokit/connection.rb:28:in `post'
	from /usr/lib/ruby/gems/2.4.0/gems/octokit-4.8.0/lib/octokit/client/statuses.rb:43:in `create_status'
	from /usr/lib/ruby/gems/2.4.0/gems/octokit-4.8.0/lib/octokit.rb:46:in `method_missing'
	from /opt/resource/lib/status.rb:15:in `create!'
	from /opt/resource/lib/commands/out.rb:59:in `block in output'
	from /opt/resource/lib/commands/out.rb:52:in `each'
	from /opt/resource/lib/commands/out.rb:52:in `output'
	from /opt/resource/lib/commands/out.rb:117:in `<main>'
```

I just rebased the master branch of the sourse and push it to this PR.

File: pkg/apis/garden/validation/validation.go [hunk 1]
```diff
@@ -39,6 +39,37 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
+type dnsSet struct {
```
Comment (timuthy) [inline pkg/apis/garden/validation/validation.go]: Would recommend to use a map directly with an empty struct, e.g. `type dnsSet map[garden.DNSProvider]struct{}`

File: pkg/apis/garden/validation/validation.go [hunk 2]
```diff
@@ -39,6 +39,37 @@ import (
...
+
+func (s *dnsSet) list() []string {
+	r := []string{}
+	for k := range s.set {
+		r = append(r, string(k))
+	}
+	return r
+}
+
+func (s *dnsSet) checkProvider(p garden.DNSProvider) bool {
```
Comment (timuthy) [inline pkg/apis/garden/validation/validation.go]: Can be omitted by using a map like proposed in the comment above.

File: pkg/apis/garden/validation/validation.go [hunk 3]
```diff
@@ -1424,12 +1455,15 @@ func validateKubernetesVersionUpdate(new, old string, fldPath *field.Path) field
...
-
-	if dns.Provider != garden.DNSUnmanaged && dns.Provider != garden.DNSAWSRoute53 && dns.Provider != garden.DNSGoogleCloudDNS && dns.Provider != garden.DNSOpenstackDesignate {
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("provider"), dns.Provider, []string{string(garden.DNSUnmanaged), string(garden.DNSAWSRoute53), string(garden.DNSGoogleCloudDNS), string(garden.DNSOpenstackDesignate)}))
+	if !availableDNS.checkProvider(dns.Provider) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("provider"), dns.Provider, availableDNS.list()))
 	}
 
-	if dns.HostedZoneID != nil {
+	if dns.Provider == garden.DNSAlicloud {
+		if dns.HostedZoneID != nil {
```
Comment (timuthy) [inline pkg/apis/garden/validation/validation.go:1445]: This conflicts with Shoot reconciliation when Shoot is set up with a customer's domain. https://github.com/gardener/gardener/blob/cb61e3ca143283691f98a8c5675e9852c1714cda/pkg/operation/botanist/dns.go#L324

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/validation/validation.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:1445]: @jia-jerry what is the reason for not allowing a hosted zone id in Alicloud DNS?

File: charts/gardener/templates/secret-internal-domain.yaml [hunk 4]
```diff
@@ -13,12 +13,17 @@ metadata:
...
     dns.garden.sapcloud.io/domain: {{ ( required ".controller.internalDomain.domain is required" .Values.controller.internalDomain.domain ) }}
+    {{- if ne $domain.provider "alicloud-dns" }}
     dns.garden.sapcloud.io/hostedZoneID: {{ ( required ".controller.internalDomain.hostedZoneID is required" .Values.controller.internalDomain.hostedZoneID ) }}
+    {{- end }}
 type: Opaque
 data:
 {{- if eq .Values.controller.internalDomain.provider "aws-route53" }}
   accessKeyID: {{ ( required ".controller.internalDomain.credentials.accessKeyID is required" .Values.controller.internalDomain.credentials.accessKeyID ) | b64enc }}
   secretAccessKey: {{ ( required ".controller.internalDomain.credentials.secretAccessKey is required" .Values.controller.internalDomain.credentials.secretAccessKey ) | b64enc }}
+{{- else if eq .Values.controller.internalDomain.provider "alicloud-dns" }}
```
Comment (rfranzke) [inline charts/gardener/templates/secret-internal-domain.yaml:24]: Can't this be merged with line 21? Let's not duplicate this.

File: charts/gardener/templates/secret-default-domain.yaml [hunk 5]
```diff
@@ -15,12 +15,17 @@ metadata:
...
     dns.garden.sapcloud.io/domain: {{ ( required ".controller.defaultDomains[].domain is required" $domain.domain ) }}
+    {{- if ne $domain.provider "alicloud-dns" }}
     dns.garden.sapcloud.io/hostedZoneID: {{ ( required ".controller.defaultDomains[].hostedZoneID is required" $domain.hostedZoneID ) }}
+    {{- end }}
 type: Opaque
 data:
 {{- if eq $domain.provider "aws-route53" }}
   accessKeyID: {{ ( required ".controller.defaultDomains[].credentials.accessKeyID is required" $domain.credentials.accessKeyID ) | b64enc }}
   secretAccessKey: {{ ( required ".controller.defaultDomains[].credentials.secretAccessKey is required" $domain.credentials.secretAccessKey ) | b64enc }}
+{{- else if eq $domain.provider "alicloud-dns" }}
```
Comment (rfranzke) [inline charts/gardener/templates/secret-default-domain.yaml:26]: Can't this be merged with line 23? Let's not duplicate this.

File: charts/gardener/values.yaml [hunk 6]
```diff
@@ -75,6 +75,10 @@ controller:
       #
       # for google-clouddns we need to provide the following keys:
       # serviceaccount.json: '{"type": "service_account", "project_id": "...", ...}'
+      # 
+      # for alicloud-dns we need to provide the following keys:
```
Comment (rfranzke) [inline charts/gardener/values.yaml:79]: Can't this be merged with line 72? Let's not duplicate this.

File: pkg/operation/garden/garden.go [hunk 7]
```diff
@@ -41,6 +42,22 @@ func New(projectLister gardenlisters.ProjectLister, namespace string) (*Garden,
...
 }
 
+//Check annotation "dns.garden.sapcloud.io/hostedZoneID", return true if the field is satisfied
+func checkDNSHostedZoneID(annotations map[string]string) bool {
+	if annotations == nil {
+		return false
+	}
+
+	r := true
+	//DNSHostedZoneID is not needed for AlicloudDNS
```
Comment (rfranzke) [inline pkg/operation/garden/garden.go]: why? please add more comments here.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/validation/validation.go (hunk 3 above)
Comment (jia-jerry) [reply pkg/apis/garden/validation/validation.go:1445]: @rfranzke There is **NO** Hosted Zone information.
1. Not found in in terraform resource structure: https://www.terraform.io/docs/providers/alicloud/d/dns_records.html
2. Not found in portal page:
https://dns.console.aliyun.com
3. Not found in API:
https://help.aliyun.com/document_detail/29772.html (In Chinese)

File: charts/gardener/templates/secret-internal-domain.yaml (hunk 4 above)
Comment (jia-jerry) [reply charts/gardener/templates/secret-internal-domain.yaml:24]: No. Content in lines 22-23 are not the same as in lines 25-26.

File: charts/gardener/templates/secret-default-domain.yaml (hunk 5 above)
Comment (jia-jerry) [reply charts/gardener/templates/secret-default-domain.yaml:26]: The content in line 25 is not the same as line 28.

File: charts/gardener/values.yaml (hunk 6 above)
Comment (jia-jerry) [reply charts/gardener/values.yaml:79]: No. For Alicloud, the naming is not the same as AWS. To track clear reference to secret which Alicloud needs, I keep the orginal name from Alicloud portal[ see https://usercenter.console.aliyun.com/#/manage/ak]

File: pkg/operation/garden/garden.go (hunk 7 above)
Comment (jia-jerry) [reply pkg/operation/garden/garden.go]: Please see previous comment. Hosted Zone is not supported by Alicloud DNS.

For all DNS providers other than Alicloud, annotation *common.DNSHostedZoneID* is a must.  For Alicloud, this annotaion is not needed.

This fuction is an extraction for one statement
```metav1.HasAnnotation(metadata, common.DNSHostedZoneID)```
because when Alicloud is envolved, the DNS provider needs to be checked together with *common.DNSHostedZoneID* annotaion.

File: pkg/apis/garden/validation/validation.go (hunk 3 above)
Comment (jia-jerry) [reply pkg/apis/garden/validation/validation.go]: @timuthy  This change does not break other DNS providers validation. Would you please check it ?

File: pkg/apis/garden/validation/validation.go (hunk 1 above)
Comment (jia-jerry) [reply pkg/apis/garden/validation/validation.go]: Great suggestion. Will follow up.

File: pkg/apis/garden/validation/validation.go (hunk 3 above)
Comment (jia-jerry) [reply pkg/apis/garden/validation/validation.go:1445]: @timuthy 
Thank you for your suggestion. Will fix the issue.
@rfranzke 
I found to add a DNS provider, it needs to take care of many switch cases statements.

Is there a plan to do some refactory? Just make it like a cloud botainist?

File: pkg/apis/garden/validation/validation.go [hunk 8]
```diff
@@ -39,6 +39,28 @@ import (
...
+
+func (s dnsSet) list() []string {
+	r := []string{}
+	for k := range s {
+		r = append(r, string(k))
+	}
+	return r
+}
+
+var availableDNS dnsSet
```
Comment (rfranzke) [inline pkg/apis/garden/validation/validation.go]: Why not using Sets (`k8s.io/apimachinery/pkg/util/sets`), already providing `Insert`, `Has` and `List` methods?

File: pkg/apis/garden/validation/validation.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:1445]: > @rfranzke There is NO Hosted Zone information.

OK, thank you.

> Is there a plan to do some refactory? Just make it like a cloud botainist?

Definitely, see GEP proposal (open PR).

File: pkg/operation/botanist/dns.go [hunk 9]
```diff
@@ -137,6 +161,11 @@ func (b *Botanist) DestroyDNSRecord(terraformerPurpose string, purposeInternalDo
 		if err != nil {
 			return err
 		}
+	case gardenv1beta1.DNSAlicloud:
+		tfvarsEnvironment, err = b.GenerateTerraformAlicCloudDNSVariablesEnvironment(purposeInternalDomain)
```
Comment (rfranzke) [inline pkg/operation/botanist/dns.go]: typo in function name

File: pkg/operation/botanist/dns.go [hunk 10]
```diff
@@ -164,9 +193,25 @@ func (b *Botanist) GenerateTerraformRoute53VariablesEnvironment(purposeInternalD
 	return common.GenerateTerraformVariablesEnvironment(secret, keyValueMap), nil
 }
 
-// GenerateTerraformCloudDNSVariablesEnvironment generates the environment containing the credentials which
+// GenerateTerraformAlicCloudDNSVariablesEnvironment generates the environment containing the credentials which
```
Comment (rfranzke) [inline pkg/operation/botanist/dns.go]: typo

File: pkg/operation/botanist/dns.go [hunk 11]
```diff
@@ -164,9 +193,25 @@ func (b *Botanist) GenerateTerraformRoute53VariablesEnvironment(purposeInternalD
...
 // are required to validate/apply/destroy the Terraform configuration. These environment must contain
 // Terraform variables which are prefixed with TF_VAR_.
+func (b *Botanist) GenerateTerraformAlicCloudDNSVariablesEnvironment(purposeInternalDomain bool) ([]map[string]interface{}, error) {
```
Comment (rfranzke) [inline pkg/operation/botanist/dns.go]: typo

File: pkg/operation/botanist/dns.go [hunk 12]
```diff
@@ -225,6 +270,31 @@ func (b *Botanist) GenerateTerraformDNSConfig(name, hostedZoneID, targetType str
...
+
+	name := secret.GetAnnotations()[common.DNSDomain]
+	if name == "" {
+		return nil, fmt.Errorf("Can't find annotaion %s in Secret %s/%s", common.DNSDomain, secret.Name, secret.Namespace)
+	}
+
+	host := strings.TrimSuffix(domain, "."+name)
+	if host == "" || host == domain {
+		return nil, fmt.Errorf("Domain %s is not a valid domain", domain)
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/dns.go]: Why do we have the above validation steps here?

The ShootDNSHostedZone admission plugin (maybe rename to ShootDNS now) should have it: https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go

File: plugin/pkg/shoot/dnshostedzone/admission.go [hunk 13]
```diff
@@ -135,6 +137,15 @@ func (d *DNSHostedZone) Admit(a admission.Attributes) error {
 		return nil
 	}
 
+	// Alicloud DNS needs no HostedZoneID
+	if shoot.Spec.DNS.Provider == garden.DNSAlicloud {
+		shoot.Spec.DNS.HostedZoneID = nil
```
Comment (rfranzke) [inline plugin/pkg/shoot/dnshostedzone/admission.go]: We should not set it explicitly to `nil` here (you already have validated in the static validation in `pkg/apis/garden/validation`) that it is `nil`.

File: plugin/pkg/shoot/dnshostedzone/admission.go [hunk 14]
```diff
@@ -197,6 +208,12 @@ func verifyHostedZoneID(shoot *garden.Shoot, secretBindingLister gardenlisters.S
 		if !accessKeyFound || !secretKeyFound {
 			return fmt.Errorf("specifying the `.spec.dns.hostedZoneID` field is only possible if the cloud provider secret or the secret referenced in .spec.dns.secretName contains credentials for AWS Route53 (%s and %s)", awsbotanist.AccessKeyID, awsbotanist.SecretAccessKey)
 		}
+	case garden.DNSAlicloud:
+		_, accessKeyFound := credentials.Data[alicloudbotanist.AccessKeyID]
+		_, accessSecretFound := credentials.Data[alicloudbotanist.AccessKeySecret]
+		if !accessKeyFound || !accessSecretFound {
+			return fmt.Errorf("specifying the `.spec.dns.hostedZoneID` field is only possible if the cloud provider secret or the secret referenced in .spec.dns.secretName contains credentials for Alicloud DNS (%s and %s)", alicloudbotanist.AccessKeyID, alicloudbotanist.AccessKeySecret)
```
Comment (rfranzke) [inline plugin/pkg/shoot/dnshostedzone/admission.go]: invalid error message? hosted zone id is not supported for alicloud

File: pkg/operation/garden/garden.go [hunk 15]
```diff
@@ -17,6 +17,7 @@ package garden
 import (
 	"fmt"
 
+	gardenapis "github.com/gardener/gardener/pkg/apis/garden"
```
Comment (rfranzke) [inline pkg/operation/garden/garden.go]: We never use `gardenapis` anywhere else, so please just use `garden`

File: pkg/operation/garden/garden.go [hunk 16]
```diff
@@ -41,6 +42,22 @@ func New(projectLister gardenlisters.ProjectLister, namespace string) (*Garden,
...
+	}
+
+	r := true
+	//DNSHostedZoneID is not needed for AlicloudDNS
+	skip := annotations[common.DNSProvider] == string(gardenapis.DNSAlicloud)
+	if !skip {
+		_, r = annotations[common.DNSHostedZoneID]
+	}
+
+	return r
```
Comment (rfranzke) [inline pkg/operation/garden/garden.go]: Refactor this function to improve readability:

```golang
func ... {
  if annotations == nil {
    return false
  }

  // Alicloud DNS does not support a hosted zone ID
  if dnsProvider, ok := annotations[common.DNSProvider]; ok && dnsProvider == string(garden.DNSAlicloud) {
    return true
  }

  _, ok := annotations[common.DNSHostedZoneID]
  return ok
}
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: plugin/pkg/shoot/dnshostedzone/admission.go (hunk 13 above)
Comment (jia-jerry) [reply plugin/pkg/shoot/dnshostedzone/admission.go]: I found some potential issues of addmission.go and will follow up.

File: pkg/operation/botanist/dns.go (hunk 12 above)
Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: It seems that we don't check secret for internal domain in admission.go.  This is why I put the check in dns.go just like other providers. Maybe it is not the best timing to change too much code in alicloud dns provider enablement. Can I just let it be currently?

File: pkg/operation/garden/garden.go (hunk 16 above)
Comment (jia-jerry) [reply pkg/operation/garden/garden.go]: Thanks for your example. I will do further refactory to make it more reasonable and readable.

File: plugin/pkg/shoot/dnshostedzone/admission.go (hunk 14 above)
Comment (jia-jerry) [reply plugin/pkg/shoot/dnshostedzone/admission.go]: It is an invalid message. I reused the the function verifyHostedZoneID() for Alicloud. It is quite misleading. I will do some refactory.

File: pkg/operation/botanist/dns.go (hunk 12 above)
Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: I took another look. The checks for the internal and default domain annotations are done here:
https://github.com/gardener/gardener/blob/master/pkg/operation/garden/garden.go#L68-L93
Please add what is needed for Alicloud there and ignore my previous comment.

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: This is not the right place to put domain name check.
This function only checks secrets. It has no context of expected domain, which is dynamic like Shoot API Server Domain.

Can we just leave it unchanged? There are some exception checks like https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L156, which should be covered by https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go. 
Maybe we need to do enhancement in another PR. I agree with you that precondition check should be covered by addmission controller. So that we will find issues at the first place and avoid to check logs after error happens.

Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: The Shoot domain is already validated in the static validation (pkg/apis/garden/validation).
This is checking the domain annotation on the default domain or internal domain secret. Why do you say it's not the right place?

The current place is not correct. We can't leave it unchanged, sorry. The check whether the domain annotation can be done when reading, and when it fails, the secret is not even considered later.

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: OK, I understand your concern.  This is a duplicated check. I also need to do some enhancement in next commit. Please check it then.

Here is my understanding domain related secrets and it is not quite related to the previous conversation. Correct me if I have some mis-understood.
1. **Internal Domain**(Only used for Shoot API Server?)
As internal domain secret has a provider and credentials, it could be verified statically. I find the verification code in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255 when shoot is being reconcied.  
Further enhancement: Can we add credential check in https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go . Because internal domain secret is a precondition for shoot cluster. Fail fast might be a better way.
2. **Default Domain**
Credential check is also in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255. 
Add this check in https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go . ?
3. **logic to get secret related to shoot domain is not consistent** https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L235 VS.
https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/dnshostedzone/admission.go#L160-L191

Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: Thank you!

1. This secret is read once during startup of GCM and it could be verified there. However, there are multiple cases where the credentials for setting up DNS might come from: Domain Secret, Cloud Provider Secret, explicitly referenced secret (`.spec.dns.secretName`). The function here https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/dns.go#L255 is getting the respective secret (which depends on the configuration). The ShootDNSHostedZone admission plugin is completely unrelated with the internal domain secret. No validation there.
1. Same as for internal domain, the default domains are read only during start-up. Validation could go there. However, in both cases it's difficult as we would have to validate based on the configured provider (route53, clouddns, ...). Can't we simply continue as it is and not put any more cloud specific logic in places where it doesn't need to be necessarily? With #370 the DNS controllers get separated and it will become much easier to validate autonomously. Take a look at #215 and what was needed to integrate openstack-designate. I see that alicloud-dns is little special due to non-existing hosted zones, however, apart from that the mentioned PR is a good "reference implementation".
3. Not sure what you mean, you didn't provide any further information. Can you elaborate? Anyway, I guess it boils down to what I wrote before: There are multiple cases where the credentials might come from.

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: For 3rd item:

The **sequences** of Secrets fetching and checking is not the same.

In admission controller, only ONE secret is checked. The sequence is **Default -> .spec.dns.secretName ->  Cloud Provider Secret.**

BTW, for Default secret, there is no credential check while the latters are checked. Why default secret is skipped. 

However, in dns.go, the sequence is **.spec.dns.secretName->Default->Cloud Provider Secret**.

Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: OK, thank's for the explanation and agreed, the order should be the same. Maybe you can refactor these parts and create a dedicated function that tells the caller which secret is to be used (default/referenced secret/cloud provider secret), call it in admission plugin and botanist.

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: Hi @rfranzke , 
I still have questions regarding the usage of default domain secret.
I think this credential should be hold by gardener landscape. 

When we try to add a Shoot's external domain DNS record, shouldn't we use credentials provided by this shoot (Gardener Tenant)? 

Maybe this is a business question regarding who pays money. I will not change the current code and will post an Issue, if it is not clarified before. How do you suggest?

Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: The default domain secret is optional and is not needed to be provided by the Gardener hoster. We have introduced it because we don't want our users to force to bring their own domain when creating shoots. However, if you want to enforce that, you don't need to create a default domain. Then the user has to bring it's own domain and specify the credentials either in the cloud provider secret, or in an extra secret and reference it in the shoot (`.spec.dns.secretName`).

Does this clarify your question?

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: And I also found, Domain is an resource of IaaS provider and it is always ownd by a user(or represented by Secret ID in AWS and Alicloud). Maybe it is not possible to be manipulated by other users (like Default domain secrect).....

Comment (jia-jerry) [reply pkg/operation/botanist/dns.go]: OK, Thanks for your kind long explanations and I'm now more clear about the secrets used by domains. Will follow up some code soon.

Comment (rfranzke) [reply pkg/operation/botanist/dns.go]: Cool. Sorry for not having proper documentation in the first place...

**Comment by rfranzke** [comment]
@jia-jerry can we review again? Or do you need to change more?

**Comment by jia-jerry** [comment]
@rfranzke , yes,  please help review it. Or do I need a squash?

**Comment by rfranzke** [comment]
Thanks @jia-jerry, will do on Monday! Yes, if no other things need to be changed/addressed I'd kindly ask you to squash. But you can wait with that until after my review.
Have a nice weekend!

File: example/10-secret-internal-domain.yaml [hunk 17]
```diff
@@ -34,10 +34,27 @@ data:
...
+#     dns.garden.sapcloud.io/provider: alicloud-dns
+#     dns.garden.sapcloud.io/domain: example.com
+# type: Opaque
+# data:
+#   accessKeyID: base64(access-key-id-with-aliclouddns-privileges)
+#   accessKeySecret: base64(access-key-secret-with-aliclouddns-privileges)
+
+# ---
+# # OpenStack Designate as dns provider
+# apiVnersion: v1
```
Comment (rfranzke) [inline example/10-secret-internal-domain.yaml]: typo

File: pkg/operation/botanist/dns.go [hunk 18]
```diff
@@ -225,6 +270,36 @@ func (b *Botanist) GenerateTerraformDNSConfig(name, hostedZoneID, targetType str
...
+		return nil, err
+	}
+	name, ok := secret.Annotations[common.DNSDomain]
+	// if the secret is not default domain secret or internal domain secret
+	if !ok {
+		// domain should be generated from shoot.spec.dns.domain
+		splits := strings.Split(domain, ".")
+		// Shoot validation promises the shoot.spec.dns.domain has at least one ".", and domain has more than one "."
+		l := len(splits)
+		name = fmt.Sprintf("%s.%s", splits[l-2], splits[l-1])
```
Comment (rfranzke) [inline pkg/operation/botanist/dns.go:289]: I understood that the `hostRecord` has to be the part of the DNS record **without** the domain, i.e. if the DNS record is `api.shoot.example.com` the `hostRecord` would be `api.shoot`, right?

Here you are cutting the last two parts, i.e. in the above example `example.com`. What happens if the domain is longer, i.e. `my.first.landscape.example.com` with a shoot domain `my-shoot.my-project.my.first.landscape.example.com`? The `hostRecord` would be `my-shoot.my-project.my.first.landscape`, does this still work/is this behaviour fine? Or should it rather be `my-shoot.my-project`?

**Review by rfranzke — COMMENTED** [review]
Looks very good 👍, one question left.

File: example/10-secret-internal-domain.yaml (hunk 17 above)
Comment (jia-jerry) [reply example/10-secret-internal-domain.yaml]: Will fix after I squash the previous commits.

File: pkg/operation/botanist/dns.go (hunk 18 above)
Comment (jia-jerry) [reply pkg/operation/botanist/dns.go:289]: @rfranzke  Good question.
 
 > I understood that the hostRecord has to be the part of the DNS record without the domain, i.e. if the DNS record is api.shoot.example.com the hostRecord would be api.shoot, right?

You are right. And I tested OK.

In the Second part, from my experience of usage of Alicloud, only top level domain (in xxx.xxx format) could be the name, because alicloud only **owns** top level domain.
In your case though domain my.first.landscape.example.com is specified in shoot DNS, it means the tenant owns the domain **example.com** in alicloud. If the generated domain is my-shoot.my-project.my.first.landscape.example.com, we still need to split into 2 parts: my-shoot.my-project.my.first.landscape and example.com.

Usually, the tenant probably has a second shoot with domain set my.second.landscape.example.com. Generated domain will be my-shoot.my-project.my.second.landscape.example.com and my code will handle the splitting logic.

In summary, we CAN'T use shoot.spec.dns.domain as name in terraform DNS record source.

Did I make it clear?

Comment (rfranzke) [reply pkg/operation/botanist/dns.go:289]: Interesting behaviour. Anyway, you are the expert, I have no experience with Alicloud. I just wanted to understand this part. So, if you say it behaves correctly then let's keep it like this. It does, right?

**Comment by jia-jerry** [comment]
@rfranzke , I squashed all the commits based on your and @timuthy 's review. Any other comments?

**Review by rfranzke — APPROVED** [review]
LGTM, thank you for your patience @jia-jerry !
